Do re-download corrupt objects with GVFS

dscho opened the pull request:
As of 9e59b38c88c, Git will loudly complain about corrupt objects.

That is fine, as long as the idea isn't to re-download locally-corrupted objects. But that's exactly what we want to do in VFS for Git. This is even tested for in the functional tests of VFS for Git, which has been identified as a regression in https://github.com/microsoft/VFSForGit/pull/1853.

dscho committed:
gvfs: allow corrupt objects to be re-downloaded

As of 9e59b38c88c (object-file: emit corruption errors when detected,
2022-12-14), Git will loudly complain about corrupt objects.

That is fine, as long as the idea isn't to re-download locally-corrupted
objects. But that's exactly what we want to do in VFS for Git via the
`read-object` hook, as per the `GitCorruptObjectTests` code
added in https://github.com/microsoft/VFSForGit/commit/2db0c030eb25 (New
features: [...] -  GVFS can now recover from corrupted git object files
[...] , 2018-02-16).

So let's support precisely that, and add a regression test that ensures
that re-downloading corrupt objects via the `read-object` hook works.

While at it, avoid the XOR operator to flip the bits, when we actually
want to make sure that they are turned off: Use the AND-NOT operator for
that purpose.

Helped-by: Matthew John Cheetham <mjcheetham@outlook.com>
Helped-by: Derrick Stolee <stolee@gmail.com>
Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

derrickstolee reviewed (dismissed):
LGTM. Is there any way we can test this without a hop to microsoft/vfsforgit?

dscho commented:
> LGTM. Is there any way we can test this without a hop to microsoft/vfsforgit?

Hmm. Valid point. The sad news is that the VFS for Git test case was introduced in microsoft/VFSforGit@2db0c030eb257beebf8e17f1c2ce72ffb166f533, which is one of those absurdly big and lamentably under-documented commits that I frequently use as examples why code changes really need to be crafted with skill.

But I realize that I even failed to target the correct commit with the `fixup!` in this PR, as there simply is no commit that actually deals with corrupt loose objects. Will reconsider.

dscho committed:
fixup! Hydrate missing loose objects in check_and_freshen()

If the `read-object` hook is not found, the code currently would fail
with the rather obscure message:

	fatal: Out of memory, strdup failed

The reason is because without a check whether the hook was found,
eventually `strvec_push()` would try to call `xstrdup()` with `cmd`
(which is `NULL` if the hook was not found) which would fail and print
that rather misleading message.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
@derrickstolee thank you for helping me be a better engineer. I _thought_ that I had a really good reproducer, but I hadn't; My initial draft of a test case reused the `corrupt_byte` function of t1060, which replaces the 10th byte by a NUL, but that would not trigger the same code path as VFS for Git's previously-failing test case, and would actually not require any fix to let the `read-object` hook re-download the corrupt object. Then I tried to replace the loose object by a 0-sized file, but same thing: The code path was lenient enough to give the `read-object` hook a job and everything worked without the fix. In the end, I painstakingly imitated the VFS for Git approach, where the loose object file is truncated by 8 bytes. That did it.

This experience of trying to find a way to trigger the same code path that does require the fix, however, kind of makes me think that https://github.com/microsoft/git/commit/9e59b38c88cdd79964cc92860906a0737bee3797, despite the claims in its commit message, is doing an incomplete job of reporting corrupted objects. Oh well. Maybe some day, when I have some free time...

In the meantime, @derrickstolee, could I solicit another review, please?

dscho commented:
Ach, darn it. macOS' `sed` strikes again.
